### PR TITLE
Min values should override max values

### LINKED ISSUE #261: Max size overrides min size, but the reverse should be true
## `taffy` version

0.2

## What you did

```rust
use taffy::prelude::*;

fn main() {
    let mut taffy = Taffy::new();
    let child = taffy.new_leaf(
        Style {
            size: Size { 
                width: Dimension::Points(2.), 
                height: Dimension::Points(2.) 
            },                
            ..Default::default()
        }
    ).unwrap();

    let parent = taffy.new_with_children(Style {
            size: Size {
                width: Dimension::Points(1.), 
                height: Dimension::Points(1.)
            },
            ..Default::default()        
        }, &[child]
    ).unwrap();

    let _ = taffy.compute_layout(parent, Size {
        width: AvailableSpace::Definite(1.),
        height: AvailableSpace::Definite(1.),
    });

    println!("{:#?}", taffy.layout(child).unwrap().size);
}
```

Output:

```rust
parent size = Size { width: 1.0, height: 1.0 }
child size = Size { width: 1.0, height: 2.0 }
```

## What went wrong

As the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/max-width) state:

> max-width overrides [width](https://developer.mozilla.org/en-US/docs/Web/CSS/width), but [min-width](https://developer.mozilla.org/en-US/docs/Web/CSS/min-width) overrides max-width

## Additional information

Migrated from https://github.com/bevyengine/bevy/issues/6737, as this is taffy's bug.

### PR DESCRIPTION
**Pull request by geom3trik**
Fixes #261

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Awesome! Can you add a simple test for this?

**Comment by alice-i-cecile** [comment]
Also: release notes here too :)

**Review by Weibye — APPROVED** [review]
Should be good to go once @alice-i-cecile's comments have been addressed 👍

**Review by TimJentzsch — CHANGES_REQUESTED** [review]
There are about 3 other implementations of `maybe_clamp` in this file, do we need to change it there too?

I also agree that we should have tests for this, one to ensure that `min_size` is higher priority than `max_size` and one that `max_size` is higher priority than `size` (for both width and height).

Maybe unit tests are enough for this, but it could also be valuable to create a test fixture with HTML for this to make sure we align with the browser.

**Comment by nicoburns** [comment]
This should also update the other `maybe_clamp` methods (for different type combinations) in that same file. I agree that we need a test for this. Annoying that we didn't already have one!

Side note: I originally tried to use the `.clamp` method from std to implement `maybe_clamp`, but this panics when min > max so is unfortunately unusable for our use case.

**Comment by TimJentzsch** [comment]
I assume making a default implementation for `maybe_clamp` is not possible due to the type signatures?

**Comment by TimJentzsch** [comment]
I'd also be in favor of including this behavior in the doc comment of the `maybe_clamp` trait definition so that we explicitly document that this is the intended behavior.

**Comment by nicoburns** [comment]
> I assume making a default implementation for maybe_clamp is not possible due to the type signatures?

Yes. The problem is that the Output type of `min` and `max` is not always the same as the input type. So we can't assume that we can chain them. It might be possible to make this work somehow, but the manual implementations were so small and simple I judged it not worth the effort.

**Comment by geom3trik** [comment]
> one to ensure that `min_size` is higher priority than `max_size` and one that `max_size` is higher priority than `size` (for both width and height).

I'm a bit confused what is meant by `max_size` having higher priority than `size`. I Checked what the web does, and when `size < max_size` then the result is `size` and when `size > max_size` the result is `max_size`, which is what I would expect but is this something different to `max_size` 'overriding' `size`?

**Comment by nicoburns** [comment]
> I Checked what the web does, and when size < max_size then the result is size and when size > max_size the result is max_size, which is what I would expect but is this something different to max_size 'overriding' size?

No, this is what is meant by max_size overriding size. In should be contrasted with the alternative possibility of size overriding max_size in which case the result would be size in all cases.

**Comment by TimJentzsch** [comment]
> > No, this is what is meant by max_size overriding size. In should be contrasted with the alternative possibility of size overriding max_size in which case the result would be size in all cases.
> 
> I thought that was the case but it seems strange that this is explicitely stated in the docs but `min-size` overriding `size` when `size < min-size` is not explicitely stated.

It's explicitely stated in the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/min-width):

> The element's width is set to the value of `min-width` whenever `min-width` is larger than [`max-width`](https://developer.mozilla.org/en-US/docs/Web/CSS/max-width) or [`width`](https://developer.mozilla.org/en-US/docs/Web/CSS/width).

Or are you talking about the taffy docs?

File: RELEASES.md [hunk 1]
```diff
@@ -11,6 +11,10 @@
 - Added `child_count`  method to `LayoutTree` for querying the number of children of a node. Required because the `children` method now returns an iterator instead of an array.
 - Added `is_childless` method to `LayoutTree` for querying whether a node has no children.
 
+### Fixes
+
+- `min_size` now overrides `max_size` which overrides `size` #261
```
Comment (TimJentzsch) [inline RELEASES.md:16]:
```suggestion
- In case of conflicts, `min_size` now overrides `max_size` which overrides `size` #261
```

**Review by TimJentzsch — APPROVED** [review]
Nice!

**Comment by geom3trik** [comment]
> Or are you talking about the taffy docs?

I was talking about this line in the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/max-width):
> max-width overrides [width](https://developer.mozilla.org/en-US/docs/Web/CSS/width), but [min-width](https://developer.mozilla.org/en-US/docs/Web/CSS/min-width) overrides max-width.

I'm just confused what it means by `max-width` overrides `width` in this case because unlike the `min-width` spec you quoted it doesn't say anything about only applying when `width` > `max-width`.

**Comment by nicoburns** [comment]
> Maybe unit tests are enough for this, but it could also be valuable to create a test fixture with HTML for this to make sure we align with the browser.

+1 for generated HTML tests for this in addition to unit tests. Process for adding new generated tests is documented here: https://github.com/DioxusLabs/taffy/blob/main/CONTRIBUTING.md. But to summarise:

- Make sure you have `chromedriver` installed.
- Add new fixtures to the `text_fixtures` directory
- Run `cargo run --package gentest` to generate the actual tests

File: RELEASES.md (hunk 1 above)
Comment (nicoburns) [inline RELEASES.md:16]: This being a regression, I feel like this ought to go into a 0.2.1 release (unless we're planning to release 0.3.0 imminently).

**Comment by TimJentzsch** [comment]
> I'm just confused what it means by `max-width` overrides `width` in this case because unlike the `min-width` spec you quoted it doesn't say anything about only applying when `width` > `max-width`.

The overrides only apply in case of conflict, i.e. `min-width > `width`, `max-width` < `width` and `min_width` > `max_width` because then you have to make an explicit decision.
I confirmed this with simple tests in the browser.

**Comment by geom3trik** [comment]
> The overrides only apply in case of conflict, i.e. `min-width` > `width`, `max-width`<`width`and`min_width`>`max_width` because then you have to make an explicit decision. I confirmed this with simple tests in the browser.

Okay I see. Then I think this is working correctly based on the tests I added.

File: RELEASES.md (hunk 1 above)
Comment (geom3trik) [reply RELEASES.md:16]: I think I would agree that perhaps this should be a 0.2.1 update. However I've also rebased this off of main, which includes changes set for 0.3.0 (currently). What would be the right process for having this in a 0.2.1 release while not including changes set for a later release? Maybe this is more of a gh question.

**Review by alice-i-cecile — APPROVED** [review]

File: RELEASES.md (hunk 1 above)
Comment (nicoburns) [reply RELEASES.md]: >  What would be the right process for having this in a 0.2.1 release while not including changes set for a later release?

I've created a `0.2.x` branch. You can rebase on top of that, and set it as the base branch for the pull request.

File: RELEASES.md [hunk 2]
```diff
@@ -5,6 +5,10 @@
 ### Fixes
 - `taffy::compute_layout` has been made public allowing Taffy to be used with custom storage (#263)
 
+### Fixes
+
```
Comment (nicoburns) [inline RELEASES.md:9]: We now have a duplicate heading here

Comment (geom3trik) [reply RELEASES.md]: Good spot, I missed that during the rebase.

**Comment by geom3trik** [comment]
Okay I've added some generated HTML tests for min-size overriding max-size, for both width and height. It seems there are already tests for min-size and max-size overriding size. Just need someone to check I did that correctly and then I think this is good to go.

**Review by nicoburns — APPROVED** [review]

**Pull request merged**